File: implementations/node/src/json-adapter.ts
// MAP v1.0 JSON-STRICT adapter
// Hand-rolled RFC 8259 parser with duplicate-key detection after escape resolution.

import {
  MapError,
  ERR_CANON_MCF, ERR_SCHEMA, ERR_TYPE, ERR_UTF8, ERR_DUP_KEY, ERR_LIMIT_SIZE,
} from "./errors";
import { MAX_CANON_BYTES } from "./constants";
import { rejectSurrogatesInString } from "./core";

// ────────── helpers ──────────

function isWS(ch: number): boolean {
  return ch === 0x20 || ch === 0x09 || ch === 0x0a || ch === 0x0d;
}

function rejectBomStrict(raw: Buffer): void {
  let i = 0;
  while (i < raw.length && isWS(raw[i])) i++;
  if (i + 3 <= raw.length && raw[i] === 0xef && raw[i + 1] === 0xbb && raw[i + 2] === 0xbf) {
    throw new MapError(ERR_SCHEMA, "BOM");
  }
}

function utf8DecodeStrict(raw: Buffer): string {
  try {
    const dec = new TextDecoder("utf-8", { fatal: true });
    return dec.decode(raw);
  } catch {
    throw new MapError(ERR_UTF8, "json utf8");
  }
}

// Number marker used to defer rejection
interface NumMarker { __num__: string }

type JsonValue = string | boolean | null | NumMarker | JsonValue[] | { [k: string]: JsonValue };

interface ParseResult {
  v: JsonValue;
  dupFound: boolean;
  surrogateFound: boolean;
}

// ────────── parser ──────────

function parseJsonStrictWithDups(raw: Buffer): ParseResult {
  if (raw.length > MAX_CANON_BYTES) throw new MapError(ERR_LIMIT_SIZE, "input too big");
  rejectBomStrict(raw);
  const text = utf8DecodeStrict(raw);
  let i = 0;
  let dupFound = false;
  let surrogateFound = false;

  function skipWS(): void {
    while (i < text.length && /\s/.test(text[i])) i++;
  }

  function expect(ch: string): void {
    skipWS();
    if (text[i] !== ch) throw new MapError(ERR_CANON_MCF, "json parse");
    i++;
  }

  function parseString(): string {
    if (text[i] !== '"') throw new MapError(ERR_CANON_MCF, "json parse");
    i++;
    let out = "";
    let closed = false;
    while (i < text.length) {
      const ch = text[i++];
      if (ch.charCodeAt(0) < 0x20) throw new MapError(ERR_CANON_MCF, "json parse");
      if (ch === '"') { closed = true; break; }
      if (ch === "\\") {
        if (i >= text.length) throw new MapError(ERR_CANON_MCF, "json parse");
        const esc = text[i++];
        if (esc === '"' || esc === "\\" || esc === "/") out += esc;
        else if (esc === "b") out += "\b";
        else if (esc === "f") out += "\f";
        else if (esc === "n") out += "\n";
        else if (esc === "r") out += "\r";
        else if (esc === "t") out += "\t";
        else if (esc === "u") {
          const hex = text.slice(i, i + 4);
          if (!/^[0-9a-fA-F]{4}$/.test(hex)) throw new MapError(ERR_CANON_MCF, "json parse");
          const code = parseInt(hex, 16);
          i += 4;
          if (code >= 0xd800 && code <= 0xdfff) {
            surrogateFound = true; // flag, do not append
          } else {
            out += String.fromCodePoint(code);
          }
        } else {
          throw new MapError(ERR_CANON_MCF, "json parse");
        }
      } else {
        out += ch;
      }
    }
    if (!closed) throw new MapError(ERR_CANON_MCF, "json parse");
    rejectSurrogatesInString(out);
    return out;
  }

  function parseValue(): JsonValue {
    skipWS();
    if (i >= text.length) throw new MapError(ERR_CANON_MCF, "json parse");
    const c = text[i];
    if (c === "{") return parseObject();
    if (c === "[") return parseArray();
    if (c === '"') return parseString();
    if (c === "t" && text.startsWith("true", i)) { i += 4; return true; }
    if (c === "f" && text.startsWith("false", i)) { i += 5; return false; }
    if (c === "n" && text.startsWith("null", i)) { i += 4; return null; }
    if (c === "-" || (c >= "0" && c <= "9")) {
      const m = text.slice(i).match(/^-?(0|[1-9]\d*)(\.\d+)?([eE][+-]?\d+)?/);
      if (!m) throw new MapError(ERR_CANON_MCF, "json parse");
      i += m[0].length;
      return { __num__: m[0] } as NumMarker;
    }
    throw new MapError(ERR_CANON_MCF, "json parse");
  }

  function parseObject(): Record<string, JsonValue> {
    const obj: Record<string, JsonValue> = {};
    const seen = new Set<string>();
    i++; // skip {
    skipWS();
    if (text[i] === "}") { i++; return obj; }
    while (true) {
      skipWS();
      const key = parseString();
      rejectSurrogatesInString(key);
      if (seen.has(key)) {
        dupFound = true;
        skipWS(); expect(":"); parseValue(); skipWS();
        if (text[i] === "}") { i++; break; }
        expect(","); continue;
      }
      seen.add(key);
      skipWS();
      expect(":");
      const val = parseValue();
      obj[key] = val;
      skipWS();
      if (text[i] === "}") { i++; break; }
      expect(",");
    }
    return obj;
  }

  function parseArray(): JsonValue[] {
    const arr: JsonValue[] = [];
    i++; // skip [
    skipWS();
    if (text[i] === "]") { i++; return arr; }
    while (true) {
      const v = parseValue();
      arr.push(v);
      skipWS();
      if (text[i] === "]") { i++; break; }
      expect(",");
    }
    return arr;
  }

  const v = parseValue();
  skipWS();
  if (i !== text.length) throw new MapError(ERR_CANON_MCF, "trailing json");
  return { v, dupFound, surrogateFound };
}

// ────────── JSON-to-canon value conversion ──────────

export function jsonToCanonValue(x: JsonValue): unknown {
  if (x && typeof x === "object" && "__num__" in x) throw new MapError(ERR_TYPE, "number");
  if (x === null) throw new MapError(ERR_TYPE, "null");
  if (typeof x === "boolean") return x ? "true" : "false";
  if (typeof x === "string") return x;
  if (Array.isArray(x)) return x.map(jsonToCanonValue);
  if (typeof x === "object") {
    const out: Record<string, unknown> = {};
    for (const k of Object.keys(x)) {
      out[k] = jsonToCanonValue(x[k]);
    }
    return out;
  }
  throw new MapError(ERR_TYPE, "unknown");
}

// ────────── public JSON adapter API ──────────

export { parseJsonStrictWithDups, ParseResult, JsonValue };

File: implementations/rust/src/json_adapter.rs
//! MAP v1.1 JSON-STRICT adapter — converts raw UTF-8 JSON bytes into the
//! canonical model.
//!
//! Type mapping (§8.2):
//!   JSON object  → MAP
//!   JSON array   → LIST
//!   JSON string  → STRING
//!   JSON boolean → BOOLEAN    (v1.1 — was STRING in v1.0)
//!   JSON integer → INTEGER    (v1.1 — was ERR_TYPE in v1.0)
//!   JSON float   → ERR_TYPE   (decimal point or exponent = rejected)
//!   JSON null    → ERR_TYPE
//!
//! The trickiest part of this module is float vs integer detection.  With
//! serde_json's `arbitrary_precision` feature, numbers arrive through the
//! deserializer as a special map with key "$serde_json::private::Number"
//! containing the raw JSON token string.  This lets us inspect for '.'
//! and 'e'/'E' directly per §8.2.1.
//!
//! Duplicate key detection requires a custom deserialization strategy since
//! serde_json's default Value type deduplicates keys silently.  Our custom
//! Deserialize impl preserves all key-value pairs via visit_map.

use serde::de::{self, Deserialize, Deserializer, MapAccess, SeqAccess, Visitor};
use std::fmt;

use crate::constants::*;
use crate::errors::*;
use crate::value::MapValue;

// The magic key serde_json uses internally to pass raw number tokens
// through serde's deserialization when arbitrary_precision is enabled.
// This is a serde_json implementation detail, but it's stable and
// well-documented in their codebase.
const SERDE_JSON_NUMBER_KEY: &str = "$serde_json::private::Number";

// ── Custom JSON value that preserves duplicate keys ────────────
// serde_json::Value uses a BTreeMap/Map which deduplicates keys.  We need
// to preserve all pairs to detect duplicates after escape resolution,
// which serde has already done for us.

#[derive(Debug)]
enum ParsedJson {
    Null,
    Bool(bool),
    /// Raw number token string (preserved by arbitrary_precision feature).
    /// Contains the exact JSON source token, e.g. "42", "3.14", "1e5".
    Number(String),
    String(String),
    Array(Vec<ParsedJson>),
    /// Preserves all key-value pairs including duplicates.
    Object(Vec<(String, ParsedJson)>),
}

struct ParsedJsonVisitor;

impl<'de> Visitor<'de> for ParsedJsonVisitor {
    type Value = ParsedJson;

    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
        formatter.write_str("any JSON value")
    }

    fn visit_bool<E: de::Error>(self, v: bool) -> Result<ParsedJson, E> {
        Ok(ParsedJson::Bool(v))
    }

    // Fallback numeric visitors — with arbitrary_precision these normally
    // won't fire (numbers route through visit_map with the magic key),
    // but we handle them defensively.
    fn visit_i64<E: de::Error>(self, v: i64) -> Result<ParsedJson, E> {
        Ok(ParsedJson::Number(v.to_string()))
    }

    fn visit_u64<E: de::Error>(self, v: u64) -> Result<ParsedJson, E> {
        Ok(ParsedJson::Number(v.to_string()))
    }

    fn visit_f64<E: de::Error>(self, v: f64) -> Result<ParsedJson, E> {
        Ok(ParsedJson::Number(v.to_string()))
    }

    fn visit_str<E: de::Error>(self, v: &str) -> Result<ParsedJson, E> {
        Ok(ParsedJson::String(v.to_string()))
    }

    fn visit_string<E: de::Error>(self, v: String) -> Result<ParsedJson, E> {
        Ok(ParsedJson::String(v))
    }

    fn visit_unit<E: de::Error>(self) -> Result<ParsedJson, E> {
        Ok(ParsedJson::Null)
    }

    fn visit_seq<A: SeqAccess<'de>>(self, mut seq: A) -> Result<ParsedJson, A::Error> {
        let mut items = Vec::new();
        while let Some(item) = seq.next_element::<ParsedJson>()? {
            items.push(item);
        }
        Ok(ParsedJson::Array(items))
    }

    fn visit_map<A: MapAccess<'de>>(self, mut map: A) -> Result<ParsedJson, A::Error> {
        // With serde_json's arbitrary_precision feature, numbers are routed
        // through visit_map as: {"$serde_json::private::Number": "raw_token"}.
        // We detect this by checking the first key.
        let first_key: Option<String> = map.next_key()?;

        match first_key {
            Some(ref key) if key == SERDE_JSON_NUMBER_KEY => {
                // This is a raw number token from serde_json's arbitrary_precision
                let raw: String = map.next_value()?;
                Ok(ParsedJson::Number(raw))
            }
            Some(first_key) => {
                // Regular JSON object — collect all pairs including duplicates
                let mut pairs = Vec::new();
                let first_value: ParsedJson = map.next_value()?;
                pairs.push((first_key, first_value));

                while let Some(key) = map.next_key::<String>()? {
                    let value: ParsedJson = map.next_value()?;
                    pairs.push((key, value));
                }
                Ok(ParsedJson::Object(pairs))
            }
            None => {
                // Empty object {}
                Ok(ParsedJson::Object(Vec::new()))
            }
        }
    }
}

impl<'de> Deserialize<'de> for ParsedJson {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        deserializer.deserialize_any(ParsedJsonVisitor)
    }
}

// ── Surrogate detection ──────────────────────────────────────
// Rust's str type guarantees valid UTF-8 (which excludes encoded surrogates),
// so in practice this check is defense-in-depth.  serde_json correctly
// rejects lone surrogates during parsing as well.

fn ensure_no_surrogates(s: &str) -> Result<(), MapError> {
    for ch in s.chars() {
        let cp = ch as u32;
        if (0xD800..=0xDFFF).contains(&cp) {
            return Err(MapError::new(
                ERR_UTF8,
                format!("surrogate U+{:04X} in JSON string", cp),
            ));
        }
    }
    Ok(())
}

/// Pre-scan raw JSON bytes for \uD800–\uDFFF escape sequences.
///
/// serde_json catches these too, but reports them as generic parse errors
/// whose message text varies across library versions.  Scanning ourselves
/// guarantees the spec-required ERR_UTF8 code regardless of serde version.
/// A high surrogate followed by a low surrogate is still rejected — JSON
/// text is UTF-8, and surrogates are only meaningful in UTF-16.
fn scan_for_surrogate_escapes(raw: &[u8]) -> Result<(), MapError> {
    let mut in_string = false;
    let mut i = 0;
    while i < raw.len() {
        let b = raw[i];
        if !in_string {
            if b == b'"' {
                in_string = true;
            }
            i += 1;
            continue;
        }
        // Inside a string.
        if b == b'\\' {
            i += 1;
            if i >= raw.len() {
                break;
            }
            if raw[i] == b'u' && i + 4 < raw.len() {
                if let Ok(hex) = std::str::from_utf8(&raw[i + 1..i + 5]) {
                    if let Ok(cp) = u16::from_str_radix(hex, 16) {
                        if cp >= 0xD800 && cp <= 0xDFFF {
                            return Err(MapError::new(
                                ERR_UTF8,
                                format!("surrogate escape \\u{}", hex),
                            ));
                        }
                    }
                }
                i += 5;
                continue;
            }
            i += 1;
            continue;
        }
        if b == b'"' {
            in_string = false;
        }
        i += 1;
    }
    Ok(())
}

// ── JSON parse with BOM and duplicate detection ────────────────

/// Parse raw JSON bytes under JSON-STRICT rules.
///
/// Returns `(parsed_value, dup_found)`.  Duplicate detection does NOT
/// short-circuit — we record the flag and keep parsing so that
/// higher-precedence errors (ERR_TYPE from null, ERR_UTF8 from bad
/// encoding) can still surface.  The caller raises ERR_DUP_KEY only
/// if no higher-precedence error occurred.
fn json_strict_parse_with_dups(raw: &[u8]) -> Result<(ParsedJson, bool), MapError> {
    if raw.len() > MAX_CANON_BYTES {
        return Err(MapError::new(
            ERR_LIMIT_SIZE,
            "input exceeds MAX_CANON_BYTES",
        ));
    }

    // BOM rejection (§8.1.1): check after skipping JSON whitespace.
    // JSON whitespace per RFC 8259: space (0x20), tab (0x09), LF (0x0A), CR (0x0D).
    let start = raw
        .iter()
        .position(|&b| b != 0x20 && b != 0x09 && b != 0x0A && b != 0x0D)
        .unwrap_or(raw.len());
    if start < raw.len() && raw[start..].starts_with(&[0xEF, 0xBB, 0xBF]) {
        return Err(MapError::new(ERR_SCHEMA, "UTF-8 BOM rejected"));
    }

    // Validate UTF-8
    let text = std::str::from_utf8(raw)
        .map_err(|_| MapError::new(ERR_UTF8, "invalid UTF-8 in JSON input"))?;

    // Pre-scan for surrogate escape sequences (§8.1).
    // serde_json rejects \uD800–\uDFFF but reports them as generic parse
    // errors whose message text varies across versions.  We scan the raw
    // bytes ourselves so the error is always ERR_UTF8 per spec precedence.
    // Same approach as the Go implementation's scanForSurrogateEscapes.
    scan_for_surrogate_escapes(raw)?;

    // Parse with our custom type that preserves duplicate keys.
    // serde_json with arbitrary_precision preserves raw number tokens.
    let parsed: ParsedJson = serde_json::from_str(text).map_err(|_| {
        MapError::new(ERR_CANON_MCF, "JSON parse error")
    })?;

    // Detect duplicates by scanning all Object nodes.
    // Duplicate detection occurs after escape resolution (§8.3) which
    // serde_json has already done for us.
    let mut dup_found = false;
    check_duplicates(&parsed, &mut dup_found)?;

    Ok((parsed, dup_found))
}

/// Recursively check for duplicate keys in all objects and validate
/// string surrogate freedom.
fn check_duplicates(val: &ParsedJson, dup_found: &mut bool) -> Result<(), MapError> {
    match val {
        ParsedJson::Object(pairs) => {
            let mut seen = std::collections::HashSet::new();
            for (key, value) in pairs {
                ensure_no_surrogates(key)?;
                if !seen.insert(key.as_str()) {
                    *dup_found = true;
                }
                check_duplicates(value, dup_found)?;
            }
        }
        ParsedJson::Array(items) => {
            for item in items {
                check_duplicates(item, dup_found)?;
            }
        }
        ParsedJson::String(s) => {
            ensure_no_surrogates(s)?;
        }
        _ => {}
    }
    Ok(())
}

// ── JSON value → canonical model ─────────────────────────────

/// Convert a parsed JSON value to a MapValue.
///
/// After this function, the value tree contains only the six canonical types.
/// Depth tracking starts at 1 for the root (which is a container).
fn json_to_canon_value(x: &ParsedJson, depth: u32) -> Result<MapValue, MapError> {
    if depth > MAX_DEPTH {
        return Err(MapError::new(ERR_LIMIT_DEPTH, "exceeds MAX_DEPTH"));
    }

    match x {
        ParsedJson::Object(pairs) => {
            // Collect unique keys (first occurrence wins, matching Python reference).
            let mut seen = std::collections::HashSet::new();
            let mut entries: Vec<(String, MapValue)> = Vec::new();

            for (key, val) in pairs {
                ensure_no_surrogates(key)?;
                if !seen.insert(key.clone()) {
                    continue; // skip duplicates (keep first)
                }
                let child_depth = match val {
                    ParsedJson::Object(_) | ParsedJson::Array(_) => depth + 1,
                    _ => depth,
                };
                let child = json_to_canon_value(val, child_depth)?;
                entries.push((key.clone(), child));
            }

            // Sort by raw UTF-8 byte order (§3.5).
            // Rust's &[u8] Ord is unsigned-byte lexicographic, which is correct.
            entries.sort_by(|(a, _), (b, _)| a.as_bytes().cmp(b.as_bytes()));

            Ok(MapValue::Map(entries))
        }

        ParsedJson::Array(items) => {
            let mut result = Vec::with_capacity(items.len());
            for item in items {
                let child_depth = match item {
                    ParsedJson::Object(_) | ParsedJson::Array(_) => depth + 1,
                    _ => depth,
                };
                result.push(json_to_canon_value(item, child_depth)?);
            }
            Ok(MapValue::List(result))
        }

        ParsedJson::String(s) => {
            ensure_no_surrogates(s)?;
            Ok(MapValue::String(s.clone()))
        }

        ParsedJson::Bool(b) => Ok(MapValue::Boolean(*b)),

        ParsedJson::Null => Err(MapError::new(ERR_TYPE, "JSON null not allowed")),

        ParsedJson::Number(token) => {
            // §8.2.1: Token-level check on the raw JSON number string.
            // Reject if contains '.' or 'e'/'E' — this is intentional.
            // Prevents silent coercion of e.g. "1.0" to integer 1.
            if token.contains('.') || token.contains('e') || token.contains('E') {
                return Err(MapError::new(
                    ERR_TYPE,
                    format!("JSON float not allowed: {}", token),
                ));
            }

            // Parse as integer and range-check against i64 bounds.
            // Use i128 to detect overflow without panicking.
            let val: i128 = token.parse().map_err(|_| {
                MapError::new(ERR_TYPE, format!("invalid integer: {}", token))
            })?;

            if val < i64::MIN as i128 || val > i64::MAX as i128 {
                return Err(MapError::new(
                    ERR_TYPE,
                    format!("integer overflow: {}", token),
                ));
            }

            Ok(MapValue::Integer(val as i64))
        }
    }
}

// ── Public JSON-STRICT API ───────────────────────────────────

/// Parse raw UTF-8 JSON bytes under JSON-STRICT rules and convert to MapValue.
///
/// Returns `(canonical_value, dup_found)`.  The caller decides whether to
/// raise ERR_DUP_KEY based on error precedence — if we made it this far
/// without a higher-precedence error, and dup_found is true, the caller
/// should reject with ERR_DUP_KEY.
pub fn parse_json_strict(raw: &[u8]) -> Result<(MapValue, bool), MapError> {
    let (parsed, dup_found) = json_strict_parse_with_dups(raw)?;
    let val = json_to_canon_value(&parsed, 1)?;
    Ok((val, dup_found))
}
